### PR TITLE
feat - show DEX quotes on LLD Swap screen

### PR DESCRIPTION
**Pull request by cjordan-ledger**
### 📝 Description

On desktop, this PR will change which rates are shown on swap - so we are now showing the DEX quotes. In the common library, there's some logic that if we search for a DEX compatible pair that we can't get a quote back for, we just show a link to the provider's iframe instead of the quote.

On mobile, we are not making any changes to this.

We only show the DEX quotes when it's behind the feature flag **swapShowDexQuotes**

**Note:**. DEX rates are only returned on staging at the mo (22-12-2022)

### ❓ Context

- **Impacted projects**:  Mobile/Desktop/Common
- **Linked resource(s)**: https://ledgerhq.atlassian.net/browse/LIVE-4633 / https://ledgerhq.atlassian.net/browse/LIVE-4862

### ✅ Checklist

- [ ] **Test coverage** 
- [ ] **Atomic delivery** 
- [ ] **No breaking changes** 

### 📸 Demo

https://user-images.githubusercontent.com/119950081/207909281-f6e231b0-3a51-43ae-8b86-d5a2f138c417.mov

#### One DEX quote

image

#### Two DEX quotes

image

#### Incompatible DEX pair 

image

### Two DEX quotes, but hidden behind feature flag (i.e. showing what we have today)

image

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: libs/ledger-live-common/src/exchange/swap/getExchangeRates.ts [hunk 1]
```diff
@@ -46,12 +46,12 @@ const getExchangeRates: GetExchangeRates = async (
   const dexProviders = ["paraswap", "oneinch"];
```
Comment (sarneijim) [inline libs/ledger-live-common/src/exchange/swap/getExchangeRates.ts:46]: Instead have this array, can we move this logic to the provider definition?
```
.filter((provider) =>
      provider.type === "DEX" ? includeDEX : true
  )

File: libs/ledger-live-common/src/exchange/swap/index.ts [hunk 2]
```diff
@@ -101,6 +101,24 @@ const swapProviders: Record<string, SwapProviderConfig> = {
...
+    needsBearerToken: false,
+  },
+  paraswap: {
+    ...createExchangeProviderNameAndSignature({
+      name: "paraswap",
+      signature: "",
+      publicKey: "",
+    }),
+    needsKYC: false,
+    needsBearerToken: false,
```
Comment (sarneijim) [inline libs/ledger-live-common/src/exchange/swap/index.ts:120]: Can we add in the  if the provider definition a new attribute, type. DEX or a CEX?

File: libs/ledger-live-common/src/exchange/swap/index.ts [hunk 3]
```diff
@@ -101,6 +101,24 @@ const swapProviders: Record<string, SwapProviderConfig> = {
...
+  },
+  paraswap: {
+    ...createExchangeProviderNameAndSignature({
+      name: "paraswap",
+      signature: "",
+      publicKey: "",
```
Comment (sarneijim) [inline libs/ledger-live-common/src/exchange/swap/index.ts:117]: Not sure if CEX attributes have sense in DEX or it is better have 2 different types (SwapDexProviderConfig, SwapCexProviderConfig, or define attribute as non-required "?") 
@jnicoulaud-ledger, could you help us? public key and signature exists or could exist in CEX context?

File: libs/ledger-live-common/src/exchange/swap/getExchangeRates.ts (hunk 1 above)
Comment (cjordan-ledger) [reply libs/ledger-live-common/src/exchange/swap/getExchangeRates.ts:46]: I've added the type to `swap/index.ts`, so I can change it to this `getProviderConfig(provider.provider).type === "DEX" ? includeDEX : true`

File: libs/ledger-live-common/src/exchange/swap/index.ts (hunk 3 above)
Comment (cjordan-ledger) [reply libs/ledger-live-common/src/exchange/swap/index.ts]: https://github.com/LedgerHQ/ledger-live/pull/2123/commits/c8822e29ecec8e1c7a4c8f34a9d212788fd50272

File: libs/ledger-live-common/src/exchange/swap/index.ts (hunk 2 above)
Comment (cjordan-ledger) [reply libs/ledger-live-common/src/exchange/swap/index.ts]: done [c8822e2](https://github.com/LedgerHQ/ledger-live/pull/2123/commits/c8822e29ecec8e1c7a4c8f34a9d212788fd50272)

Comment (sarneijim) [reply libs/ledger-live-common/src/exchange/swap/index.ts]: Not needed

**Comment by sarneijim** [comment]
Can you add the correct description to the pr? and add changelogs?

**Comment by cjordan-ledger** [comment]
> Can you add the correct description to the pr? and add changelogs?

@sarneijim I've added in changelog/included the JIRA ticket. I have already put in a description, let me know if there's anything I should change.

**Comment by sarneijim** [comment]
> > Can you add the correct description to the pr? and add changelogs?
> 
> @sarneijim I've added in changelog/included the JIRA ticket. I have already put in a description, let me know if there's anything I should change.

Could you use one changelog per package? (ci has some problem to detect when they are in the same file)

**Comment by cjordan-ledger** [comment]
> > > Can you add the correct description to the pr? and add changelogs?
> > 
> > 
> > @sarneijim I've added in changelog/included the JIRA ticket. I have already put in a description, let me know if there's anything I should change.
> 
> Could you use one changelog per package? (ci has some problem to detect when they are in the same file)

@sarneijim updated: https://github.com/LedgerHQ/ledger-live/pull/2123/commits/f6b47d76e30ec1401ea15a8919e7edc9ad192c20

**Review by sarneijim — APPROVED** [review]

File: libs/ledger-live-common/src/exchange/swap/completeExchange.ts [hunk 4]
```diff
@@ -73,6 +73,10 @@ const completeExchange = (
         const errorsKeys = Object.keys(errors);
         if (errorsKeys.length > 0) throw errors[errorsKeys[0]]; // throw the first error
 
+        if (providerConfig.type !== "CEX") {
+          throw new Error(`Unsupported provider type ${providerConfig.type}`);
```
Comment (ggilchrist-ledger) [inline libs/ledger-live-common/src/exchange/swap/completeExchange.ts:77]: Any idea how I can enter this state and test it @cjordan-ledger @sarneijim ?

File: libs/ledger-live-common/src/exchange/swap/index.ts [hunk 5]
```diff
@@ -53,7 +53,7 @@ const getSwapAPIVersion: () => number = () => {
   return version;
 };
 
-const ftx: SwapProviderConfig = {
+const ftx: ProviderConfig = {
   ...createExchangeProviderNameAndSignature({
     name: "FTX",
```
Comment (ggilchrist-ledger) [inline libs/ledger-live-common/src/exchange/swap/index.ts:58]: Can we remove FTX stuff? @cjordan-ledger @sarneijim . Looping in @chabroA @JunichiSugiura for context

File: libs/ledger-live-common/src/exchange/swap/completeExchange.ts (hunk 4 above)
Comment (cjordan-ledger) [reply libs/ledger-live-common/src/exchange/swap/completeExchange.ts:77]: @ggilchrist-ledger AFAIK I don't believe it's currently possible to get to this line in the app, as DEX providers won't call this function. This has been added to ensure the code is type safe.

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/index.jsx [hunk 6]
```diff
@@ -395,8 +399,33 @@ const SwapForm = () => {
...
+    if (showDexQuotes) {
+      return false;
+    }
+
+    if (sourceAccount && targetAccount) {
+      const sourceMainAccount = getMainAccount(sourceAccount, sourceParentAccount);
+      const targetMainAccount = getMainAccount(targetAccount, targetParentAccount);
+
+      if (
+        targetMainAccount.currency.family === "ethereum" &&
```
Comment (sarneijim) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/index.jsx:420]: We can move this logic to common when we update llm, but could we support binance and polygon family here too?

const dexFamilyList = ["ethereum", "binance", "polygon"];

Comment (cjordan-ledger) [reply apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/index.jsx]: Good call https://github.com/LedgerHQ/ledger-live/pull/2123/commits/27c406c07719c84d236fd8860041b30b9d5d578a

File: apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/index.jsx [hunk 7]
```diff
@@ -395,8 +399,33 @@ const SwapForm = () => {
 
   const sourceAccount = swapTransaction.swap.from.account;
   const sourceCurrency = swapTransaction.swap.from.currency;
+  const sourceParentAccount = swapTransaction.swap.from.parentAccount;
+  const targetAccount = swapTransaction.swap.to.account;
+  const targetParentAccount = swapTransaction.swap.to.parentAccount;
   const targetCurrency = swapTransaction.swap.to.currency;
 
+  // We check if a decentralized swap is available to conditionnaly render an Alert below.
+  // All Ethereum related currencies are considered available
```
Comment (sarneijim) [inline apps/ledger-live-desktop/src/renderer/screens/exchange/Swap2/Form/index.jsx:408]: Binance, polygon

**Review by sarneijim — APPROVED** [review]

File: libs/ledger-live-common/src/exchange/swap/index.ts [hunk 8]
```diff
@@ -63,9 +63,18 @@ const ftx: SwapProviderConfig = {
   }),
   needsKYC: true,
   needsBearerToken: true,
+  type: "CEX",
 };
 
-const swapProviders: Record<string, SwapProviderConfig> = {
+type CEXProviderConfig = SwapProviderConfig & { type: "CEX" };
+type DEXProviderConfig = {
+  needsKYC: boolean;
```
Comment (ggilchrist-ledger) [inline libs/ledger-live-common/src/exchange/swap/index.ts:71]: Check if we need these @cjordan-ledger

**Comment by ggilchrist-ledger** [comment]
Test feedback:
- Change 'Oneinch' to '1inch'
- Investigate wrong Ethereum account selected in 1inch/paraswap Dapp
- Add missing unit tests

File: libs/ledger-live-common/src/exchange/swap/index.ts (hunk 8 above)
Comment (cjordan-ledger) [reply libs/ledger-live-common/src/exchange/swap/index.ts:71]: I think we do need these properties on DEX providers as well but they can be set to false. It makes the code simpler if all providers have `needsKYC`/`needsBearerToken`

**Comment by gre** [comment]
@cjordan-ledger in future, please use "Squash & Merge" to not create a lot of commits in develop branch. Thanks 👍

**Pull request merged**